pagefaultgames/pokerogue
# [Move] Implement Heal Block

Old PRs: #3326 #3047 

## What are the changes the user will see?
Heal Block will be implemented. 
Psychic Noise will have its secondary effect implemented - Heal Block will be applied to the opponent for two turns. 

## Why am I making these changes?
Attempt to get rid of partial tags. 

## What are the changes from a developer perspective?

HealBlockTag prevents healing in HealPhase. The Regenerator ability is unaffected by Heal Block because it does not use Heal Phase for its healing. but also extends upon the MoveRestrictionBattlerTag to restrict the use of healing moves while the tag is still in effect. The move restriction is based on if the move is a status move and has a TRIAGE_MOVE flag. This avoids adding another redundant flag since attacking moves that heal are not banned with Heal Block. In addition, Heal Block-exempt moves like Pain Split do not have the Triage Flag. There is one exception to this condition however - Pollen Puff.

Pollen Puff has two different effects depending on the move's target. If it targets an opponent, it is a 90 BP attack. If it targets an ally, it is a healing move. Heal Block prevents the affected from using Heal Block towards allies, but allows for it to be used against opponents. This means that it cannot be disabled in Command Phase, like most moves, and must be disabled in Select Target Phase. MoveRestrictionBattlerTag was created based on preventing the use of moves in Command Phase, not Select Target Phase, so I added new functions and parameters to check that Pollen Puff had an allowed use case. Pollen Puff has an attribute StatusCategoryOnAllyAttr but because it was never applied in-game until now, it didn't work. This attribute was fixed to check if the user and the target are allies. Arguably, these new additions should help with Taunt's implementation. 

Psychic Noise and Heal Block share similar implementations. However, Psychic Noise is allowed to hit the user again even when Heal Block is active while Heal Block (the move) should fail when the user already has an Heal Block tag. 

### Screenshots/Videos
Pollen Puff

https://github.com/user-attachments/assets/30824664-95f4-498d-933e-ba250bc70dbd



## How to test the changes?
Use heal block on a Pokemon and damage it afterwards to see if it can heal or not.

## Checklist
